server/commands/connected_objects.go

```diff
@@ -0,0 +1,351 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	serverErrors "github.com/openfga/openfga/server/errors"
+
+	"github.com/go-errors/errors"
+	"github.com/openfga/openfga/internal/graph"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+	"github.com/openfga/openfga/storage"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"golang.org/x/sync/errgroup"
+)
+
+type ConnectedObjectsRequest struct {
+	StoreID          string
+	ObjectType       string
+	Relation         string
+	User             *openfgapb.ObjectRelation
+	ContextualTuples []*openfgapb.TupleKey
+
+	// Limit limits the results yielded by the ConnectedObjects API.
+	Limit uint32
+}
+
+type ConnectedObjectsCommand struct {
+	Datastore        storage.OpenFGADatastore
+	Typesystem       *typesystem.TypeSystem
+	ResolveNodeLimit uint32
+}
+
+func (c *ConnectedObjectsCommand) streamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	depth, ok := graph.ResolutionDepthFromContext(ctx)
+	if !ok {
+		ctx = graph.ContextWithResolutionDepth(ctx, c.ResolveNodeLimit)
+	} else {
+		if depth == 0 {
+			return serverErrors.AuthorizationModelResolutionTooComplex
+		}
+
+		ctx = graph.ContextWithResolutionDepth(ctx, depth-1)
+	}
+
+	storeID := req.StoreID
+
+	targetUserType, _ := tuple.SplitObject(req.User.Object)
+
+	targetUserRef := &openfgapb.RelationReference{
+		Type:     targetUserType,
+		Relation: req.User.Relation,
+	}
+
+	sourceObjRef := &openfgapb.RelationReference{
+		Type:     req.ObjectType,
+		Relation: req.Relation,
+	}
+
+	// build the graph of possible edges between object types in the graph based on the authz model's type info
+	g := graph.BuildConnectedObjectGraph(c.Typesystem)
+
+	// find the possible incoming edges (ingresses) between the target user reference and the source (object, relation) reference
+	ingresses, err := g.RelationshipIngresses(sourceObjRef, targetUserRef)
+	if err != nil {
+		return err
```

Comment:
- [ ]: make sure we're surfacing errors from the `server/errors` package here and elsewhere.